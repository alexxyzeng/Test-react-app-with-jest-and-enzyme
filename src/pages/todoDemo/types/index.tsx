export interface todo {
  text: string;
}
