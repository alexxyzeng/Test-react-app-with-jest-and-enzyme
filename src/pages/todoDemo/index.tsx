import * as React from 'react';

import { connect } from 'dva';
import { todo } from './types';

interface TodoProps {
  removeTodo?: (index: number) => void;
  addTodo?: (text: string) => void;
  todos: Array<todo>;
}

interface TodoState {
  input: string;
}

class TodoList extends React.Component<TodoProps, TodoState> {
  constructor(props: TodoProps) {
    super(props);
    this.state = { input: '' };
  }

  onInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ input: event.target.value });
  };

  onTodoClick = (index: number) => () => {
    const { removeTodo } = this.props;
    if (removeTodo === undefined) {
      return;
    }
    removeTodo(index);
  };

  onTodoSubmit = () => {
    const { addTodo } = this.props;
    const { input } = this.state;
    if (addTodo === undefined) {
      return;
    }
    addTodo(input);
    this.setState({ input: '' });
  };

  render() {
    const { input } = this.state;
    const { todos } = this.props;
    return (
      <div className="todo-container">
        <h1>Todos</h1>
        <input type="text" onChange={this.onInputChange} value={input} />
        {
          <ul>
            {todos.map(({ text }, index) => {
              return (
                <li key={index} onClick={this.onTodoClick(index)}>
                  {text}
                </li>
              );
            })}
          </ul>
        }
        <button className="todo-add" onClick={this.onTodoSubmit}>
          Add Todo
        </button>
      </div>
    );
  }
}

// const mapStateProps = ({})
interface StoreState {
  todos: {
    todoList: [todo];
  };
}
const mapStateToProps = ({ todos }: StoreState) => {
  const { todoList } = todos;
  return {
    todos: todoList,
  };
};

const ADD_TODO = 'todos/addTodo';
const REMOVE_TODO = 'todos/removeTodo';

const mapDispatchToProps = (dispatch: React.Dispatch<any>) => {
  return {
    addTodo: (text: string) => dispatch({ type: ADD_TODO, payload: { text } }),
    removeTodo: (index: number) => dispatch({ type: REMOVE_TODO, payload: { index } }),
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(TodoList);

export { TodoList };
