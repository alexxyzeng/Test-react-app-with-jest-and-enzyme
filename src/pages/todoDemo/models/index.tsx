import { Model } from 'dva';
import { AnyAction } from 'redux';
import { todo } from '../types';

interface State {
  todoList: [todo];
}

const todos: Model = {
  namespace: 'todos',
  state: {
    todoList: [],
  },
  reducers: {
    addTodo(state: State, { payload: { text } }) {
      const { todoList } = state;
      return {
        todoList: [...todoList, { text }],
      };
    },
    removeTodo(state: State, { payload: { index } }) {
      const { todoList } = state;
      return {
        todoList: todoList.filter((_todo, todoIndex) => {
          return todoIndex !== index;
        }),
      };
    },
  },
};

export default todos;
