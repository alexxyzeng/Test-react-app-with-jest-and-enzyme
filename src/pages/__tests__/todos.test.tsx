import React from 'react';
import 'jest';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { TodoList } from '../todoDemo';
import { todo } from '../todoDemo/types';

configure({ adapter: new Adapter() });

describe('<TodoList />', () => {
  it('matches snapshot', () => {
    const todos: Array<todo> = [];
    const wrapper = shallow(<TodoList todos={todos} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('calls setState after input change', () => {
    const wrapper = shallow(<TodoList todos={[]} />);
    wrapper.find('input').simulate('change', { target: { value: 'Add Todo' } });
    expect(wrapper.state('input')).toEqual('Add Todo');
  });
  it('calls addTodo with submit button click', () => {
    const addTodo = jest.fn();
    const todos: Array<todo> = [];
    const wrapper = shallow(<TodoList todos={todos} addTodo={addTodo} />);
    wrapper.find('input').simulate('change', { target: { value: 'Add Todo' } });
    wrapper.find('.todo-add').simulate('click');
    expect(addTodo).toHaveBeenCalledWith('Add Todo');
  });
  it('calls removeTodo with todo item click', () => {
    const removeTodo = jest.fn();
    const todos: Array<todo> = [{ text: 'Learn Jest' }, { text: 'Learn RxJS' }];
    const wrapper = shallow(<TodoList todos={todos} removeTodo={removeTodo} />);
    wrapper
      .find('li')
      .at(0)
      .simulate('click');
    expect(removeTodo).toHaveBeenCalledWith(0);
  });
});
