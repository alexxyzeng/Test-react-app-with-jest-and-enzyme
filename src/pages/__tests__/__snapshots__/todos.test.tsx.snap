// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> matches snapshot 1`] = `
<div
  className="todo-container"
>
  <h1>
    Todos
  </h1>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  <ul />
  <button
    className="todo-add"
    onClick={[Function]}
  >
    Add Todo
  </button>
</div>
`;
